File: sdk/python/tap/chain/discriminator.py
"""Anchor instruction discriminators.

Anchor identifies an instruction by `sha256("global:<name>")[:8]`. We
hard-code the eight-byte prefixes here so the SDK doesn't take a dependency
on a SHA implementation at import time."""

from __future__ import annotations

import hashlib


def _disc(name: str) -> bytes:
    return hashlib.sha256(f"global:{name}".encode("utf-8")).digest()[:8]


OPEN_CHANNEL = _disc("open_channel")
SETTLE = _disc("settle")
DISPUTE = _disc("dispute")
CLOSE = _disc("close")

File: sdk/typescript/src/chain/discriminator.ts
/**
 * Anchor instruction discriminators. Anchor identifies an instruction by
 * `sha256("global:<name>")[:8]`. Hard-coding these would risk drift on a
 * rename; we derive them once at import time so the source-of-truth lives
 * alongside the Anchor names.
 */

import { sha256 } from "@noble/hashes/sha256";

function disc(name: string): Uint8Array {
  return sha256(new TextEncoder().encode(`global:${name}`)).slice(0, 8);
}

export const OPEN_CHANNEL = disc("open_channel");
export const SETTLE = disc("settle");
export const DISPUTE = disc("dispute");
export const CLOSE = disc("close");
